### PR TITLE
Persisting received infotons when token updates #466

### PR DESCRIPTION
**Pull request by nicklester**

### CODE REVIEW

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/TokenReporter.scala [hunk 1]
```diff
@@ -160,8 +172,10 @@ class WebExporter(reporter: ActorRef, port: Int = 8080)(implicit system: ActorSy
   }
 }
 
+case object RequestPreviousStatistics
```
Comment (e-orz) [inline server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/TokenReporter.scala]: dead code. should be removed

File: server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/StpUtil.scala [hunk 2]
```diff
@@ -1,36 +1,42 @@
...
+                DownloadStats(receivedInfotons=value)
+              }
+
+              val sensor = extractLastPart(json.hcursor.downField("system").get[String]("path").toOption.get)
+              sensor -> (token,receivedInfotons)
+            }
           }
-          .foldLeft(Map.empty[String, Token])(_ + _)
-      )
+        }).foldLeft(Map.empty[String,TokenAndStatistics])(_ + _)
```
Comment (e-orz) [inline server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/StpUtil.scala]: use:
`.foldLeft(Map.newBuilder[String,TokenAndStatistics])(_.+=(_)).result()`
instead so you won't build immutable Map in memory for each iteration....

Comment (nicklester) [reply server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/StpUtil.scala]: Addressed this...

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/TokenReporter.scala (hunk 1 above)
Comment (nicklester) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/TokenReporter.scala]: Done

File: server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/InfotonReporter.scala [hunk 3]
```diff
@@ -123,19 +125,25 @@ class InfotonReporter private(baseUrl: String, path: String)(implicit mat: Mater
...
       HttpRequest(uri = s"http://$host:$port/_in?format=$format&replace-mode", method = HttpMethods.POST, entity = data)
         .addHeader(RawHeader("X-CM-WELL-TOKEN", writeToken))
     }
 
-    def createTriple(sensor: String, token: Token) = {
+    def createTriples(sensor: String, token: Token, downloadStats: Option[DownloadStats]) = {
       val p = if (path startsWith "/") path.tail else path
-      s"""<cmwell://$p/tokens/$sensor> <cmwell://meta/nn#token> "$token" ."""
+
+      downloadStats.map { s =>
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/InfotonReporter.scala]: Part 2 Done.

File: server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/InfotonReporter.scala [hunk 4]
```diff
@@ -66,15 +65,18 @@ class InfotonReporter private(baseUrl: String, path: String)(implicit mat: Mater
     case RequestPreviousTokens =>
       context.become(receiveBeforeInitializes(sender() :: recipients))
 
-    case Success(savedTokens: Map[String, Token]) =>
+    case RequestPreviousStatistics =>
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/InfotonReporter.scala]: Done

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/stats/DownloaderStats.scala [hunk 5]
```diff
@@ -87,6 +91,10 @@ class DownloaderStats(isStderr: Boolean,
       val formatter = java.text.NumberFormat.getNumberInstance
 
       override def preStart(): Unit = {
+
+        // Initialise persisted statistics
+        initialDownloadStats.map { stats=> totalReceivedInfotons mark stats.receivedInfotons }
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/stats/DownloaderStats.scala]: Done

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/SparqlTriggeredProcessor.scala [hunk 6]
```diff
@@ -84,25 +81,33 @@ class SparqlTriggeredProcessor(config: Config,
 
   def listen()(implicit system: ActorSystem, mat: Materializer, ec: ExecutionContext) = {
 
-    def addStatsToSource(id: String, source: Source[(ByteString, Option[SensorContext]), _]) = {
-      source.via(DownloaderStats(format = "ntriples", label = Some(id), reporter = tokenReporter))
+    def addStatsToSource(id: String, source: Source[(ByteString, Option[SensorContext]), _], initialDownloadStats: Option[TokenAndStatisticsMap] = None) = {
+      source.via(DownloaderStats(format = "ntriples", label = Some(id), reporter = tokenReporter, initialDownloadStats = {
+        initialDownloadStats.map { stats =>
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/SparqlTriggeredProcessor.scala]: Done

File: server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/StpUtil.scala [hunk 7]
```diff
@@ -1,36 +1,42 @@
 package cmwell.tools.data.sparql
 
-import cmwell.tools.data.downloader.consumer.Downloader.Token
-
+import cmwell.tools.data.utils.akka.stats.DownloaderStats.DownloadStats
+import cmwell.util.http.SimpleResponse.Implicits.UTF8StringHandler
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/StpUtil.scala]: Agree. Done

**Pull request merged**